main.rb

```diff
@@ -0,0 +1,156 @@
+module Enumerable
+  def my_each
+    return to_enum :my_each unless block_given?
+
+    i = 0
+    a = is_a?(Range) ? to_a : self
+    while i < a.length
+      yield self[i]
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    i = 0
+    return to_enum :my_each unless block_given?
+
+    a = is_a?(Range) ? to_a : self
+    while i < a.length
+      yield(a[i], i)
+      i += 1
+    end
+    a
+  end
+
+  def my_select
+    return to_enum :my_each unless block_given?
+
+    arr = []
+    my_each { |x| arr.push(x) if yield(x) }
+    arr
+  end
+
+  def my_all?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == false }
+    elsif arg.nil? == false
+      my_each { |x| return false if arg != x }
+    elsif arg.is_a?(Regex)
+      my_each { |x| return false if arg.match?(x.to_s) == false }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == false }
+    elsif block_given? == false
+      to_enum
+    else
+      my_each { |x| return false unless x }
+    end
+    true
+  end
+
+  def my_any?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return true if prc.call(x) }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return true if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return true if x.is_a?(arg) }
+    elsif arg.nil? == false
+      my_each { |x| return true if arg == x }
+    elsif block_given? == false
+      to_enum
+    else
+      my_each { |x| return true if x }
+    end
+    false
+  end
+
+  def my_none?(arg = nil)
+    if block_given?
+      my_each { |x| return false if prc.call(x) }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) }
+    elsif arg.nil? == false
+      my_each { |x| return false if x == arg }
+    else
+      my_each { |x| return false if x }
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    count = 0
+    if block_given?
+      my_each { |x| count += 1 if yield(x) }
+    elsif arg.nil? == false
+      my_each { |x| count += 1 if x == arg }
+    elsif arg.nil? and block_given? == false
+      my_each { |x| count += 1 unless x.nil? }
+    end
+    count
+  end
+
+  def my_map(&proc)
+    return to_enum :my_each unless block_given?
+
+    array = []
+    if proc
+      my_each { |x| array << proc.call(x) }
+    else
+      my_each { |x| array << yield(x) }
+    end
+    array
+  end
```

Comment:
- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so let's work on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```